### PR TITLE
PatternMatch: migrate to CmpPredicate

### PR DESCRIPTION
**Pull request by artagnon**
With the introduction of CmpPredicate in 51a895a (IR: introduce struct with CmpInst::Predicate and samesign), PatternMatch is one of the first key pieces of infrastructure that must be updated to match a CmpInst respecting samesign information. Implement this change to Cmp-matchers.

This is a preparatory step in migrating the codebase over to CmpPredicate. Since we no functional changes are desired at this stage, we have chosen not to migrate CmpPredicate::operator==(CmpPredicate) calls to use CmpPredicate::getMatching(), as that would have visible impact on tests that are not yet written: instead, we call CmpPredicate::operator==(Predicate), preserving the old behavior, while also inserting a few FIXME comments for follow-ups.

### CODE REVIEW

**Comment by artagnon** [comment]
This is going to have a sizable compile-time impact, but again, I'm not sure we can help it.

**Comment by nikic** [comment]
> This is going to have a sizable compile-time impact, but again, I'm not sure we can help it.

Can you share what the compile-time impact is?

We could just provide two overloads in PatternMatch. There isn't really a need to migrate *everything* to CmpPredicate if it's not free, I expect only a small fraction of code will actually care about samesign (the bits interacting with implied conditions).

**Comment by artagnon** [comment]
> > This is going to have a sizable compile-time impact, but again, I'm not sure we can help it.
> 
> Can you share what the compile-time impact is?
> 
> We could just provide two overloads in PatternMatch. There isn't really a need to migrate _everything_ to CmpPredicate if it's not free, I expect only a small fraction of code will actually care about samesign (the bits interacting with implied conditions).

Okay, I was wrong, and it's essentially free.

https://llvm-compile-time-tracker.com/compare.php?from=c7babfa6a375d1094cf5e5d7ce4b36b2b9a6b717&to=61823cdd5e86b8ae0c45da03ce6d8faf32b86038&stat=instructions:u

**Comment by artagnon** [comment]
Gentle ping.

File: llvm/lib/IR/Instructions.cpp [hunk 1]
```diff
@@ -3932,6 +3933,26 @@ std::optional<CmpPredicate> CmpPredicate::getMatching(CmpPredicate A,
...
+CmpPredicate CmpPredicate::get(const CmpInst *Cmp) {
+  return TypeSwitch<const CmpInst *, CmpPredicate>(Cmp)
+      .Case<ICmpInst>([](auto *ICI) { return ICI->getCmpPredicate(); })
+      .Case<FCmpInst>([](auto *FCI) { return FCI->getPredicate(); });
+}
+
+CmpPredicate CmpPredicate::getSwapped(CmpPredicate P) {
+  return CmpInst::isIntPredicate(P)
+             ? ICmpInst::getSwappedCmpPredicate(P)
+             : CmpPredicate{CmpInst::getSwappedPredicate(P)};
```
Comment (nikic) [inline llvm/lib/IR/Instructions.cpp:3945]:
```suggestion
  return {CmpInst::getSwappedPredicate(P), P.SameSign};
```
Maybe? I don't think we need the branch here.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**